receiver/googlecloudspannerreceiver/internal/filter/itemcardinality_test.go

```diff
@@ -0,0 +1,334 @@
+				itemActivityPeriod, logger)
+
+			if testCase.expectError {
+				require.Nil(t, filter)
+				require.Error(t, err)
+			} else {
+				require.NoError(t, err)
+				filterCasted := filter.(*itemCardinalityFilter)
+				defer executeShutdown(t, filterCasted)
+
+				assert.Equal(t, metricName, filterCasted.metricName)
+				assert.Equal(t, testCase.totalLimit, filterCasted.totalLimit)
+				assert.Equal(t, testCase.limitByTimestamp, filterCasted.limitByTimestamp)
+				assert.Equal(t, itemActivityPeriod, filterCasted.itemActivityPeriod)
+				assert.Equal(t, logger, filterCasted.logger)
+				require.NotNil(t, filterCasted.cache)
+			}
+		})
+	}
+}
+
+func TestItemCardinalityFilter_TotalLimit(t *testing.T) {
+	itemFilter := &itemCardinalityFilter{totalLimit: totalLimit}
+
+	assert.Equal(t, totalLimit, itemFilter.TotalLimit())
+}
+
+func TestItemCardinalityFilter_LimitByTimestamp(t *testing.T) {
+	itemFilter := &itemCardinalityFilter{limitByTimestamp: limitByTimestamp}
+
+	assert.Equal(t, limitByTimestamp, itemFilter.LimitByTimestamp())
+}
+
+func TestItemCardinalityFilter_CanIncludeNewItem(t *testing.T) {
+	logger := zap.NewNop()
+	testCases := map[string]struct {
+		totalLimit         int
+		limitByTimestamp   int
+		keysAlreadyInCache []string
+		expectedResult     bool
+	}{
+		"No items in cache and timestamp limit hasn't been exhausted": {2, 1, nil, true},
+		"Cache is full and timestamp limit hasn't been exhausted":     {2, 1, []string{"qwerty1", "qwerty2"}, false},
+		"No items in cache and timestamp limit has been exhausted":    {2, 0, nil, false},
+	}
+
+	for name, testCase := range testCases {
+		t.Run(name, func(t *testing.T) {
+			filter, err := NewItemCardinalityFilter(metricName, testCase.totalLimit, testCase.limitByTimestamp,
+				itemActivityPeriod, logger)
+			require.NoError(t, err)
+			filterCasted := filter.(*itemCardinalityFilter)
+			defer executeShutdown(t, filterCasted)
+
+			for _, key := range testCase.keysAlreadyInCache {
+				err = filterCasted.cache.Set(key, byte(1))
+				require.NoError(t, err)
+			}
+
+			assert.Equal(t, testCase.expectedResult, filterCasted.canIncludeNewItem(testCase.limitByTimestamp))
+		})
+	}
+}
+
+func TestItemCardinalityFilter_Shutdown(t *testing.T) {
+	logger := zap.NewNop()
+	testCases := map[string]struct {
+		closeCache  bool
+		expectError bool
+	}{
+		"Happy path":            {false, false},
+		"Cache has been closed": {true, true},
+	}
+
+	for name, testCase := range testCases {
+		t.Run(name, func(t *testing.T) {
+			filter, err := NewItemCardinalityFilter(metricName, totalLimit, limitByTimestamp, itemActivityPeriod, logger)
+			require.NoError(t, err)
+			filterCasted := filter.(*itemCardinalityFilter)
+
+			if testCase.closeCache {
+				// Covering case when by some reasons cache is closed
+				err = filterCasted.cache.Close()
+				require.NoError(t, err)
+			}
+
+			if testCase.expectError {
```

Comment:
You can use `assert.ErrorIs(t, testcase.expectError, filter.Shutdown())` instead of using the if else test cases

Reply:
I also suggest not using require here but use assert to allow for all the test cases to proceed and capture all the errors.

Reply:
I didn't get the idea with `assert.ErrorIs()`. Also according to https://stackoverflow.com/questions/42035104/how-to-unit-test-go-errors it seems that in this case checking if error is nil/not nil is more than enough.

Reply:
It just removes the need for the if else statement here, it is done for you :)

Reply:
Aaa, sorry. brain hadn't fully kicked in.

This is fine, I am just so use to seeing `expectError` as `error` not as a `bool`

Reply:
NP :)